File: src/typescript/domainObjects/hold/HoldTicket.ts
import moment from "moment";

export default class HoldTicket {

    closedBy: string;
    closedDate: any;
    color: string;
    comments: string;
    cparNum: string;
    defectId: number;
    dispositionBy: string;
    dispositionComments: string;
    dispositionDate: any;
    dispositionId: number;
    holdDate: any;
    holdTicketId: number;
    lineItem: string;
    materialDesc: string;
    materialNumber: string;
    onHoldBy: string;
    optionNumber: string;
    parentId: number;
    plantCode: string;
    prodOrderNum: string;
    prodOrderQty: string;
    producedByShift: string;
    productionDate: any;
    purchaseOrderNum: string;
    qtyOnHold: string;
    readonly: boolean;
    responsibleForAction: string;
    returnAuthNum: string;
    reverificationRequired: string;
    salesOrderNum: string;
    skidNumbersOnHold: string;
    statusId: number;
    totalSkids: string;
    workCenter: string;


    constructor(holdTicket) {

        this.holdTicketId = holdTicket.holdTicketId;
        this.parentId = holdTicket.parentId;
        this.dispositionId = holdTicket.dispositionId;
        this.defectId = holdTicket.defectId;
        this.plantCode = holdTicket.plantCode;
        this.statusId = holdTicket.statusId;
        this.prodOrderNum = holdTicket.prodOrderNum;
        this.prodOrderQty = holdTicket.prodOrderQty;
        this.salesOrderNum = holdTicket.salesOrderNum;
        this.returnAuthNum = holdTicket.returnAuthNum;
        this.purchaseOrderNum = holdTicket.purchaseOrderNum;
        this.materialDesc = holdTicket.materialDesc;
        this.materialNumber = holdTicket.materialNumber;
        this.lineItem = holdTicket.lineItem;
        this.optionNumber = holdTicket.optionNumber;
        this.color = holdTicket.color;
        this.workCenter = holdTicket.workCenter;
        this.producedByShift = holdTicket.producedByShift;
        this.qtyOnHold = holdTicket.qtyOnHold;
        this.totalSkids = holdTicket.totalSkids;
        this.skidNumbersOnHold = holdTicket.skidNumbersOnHold;
        this.onHoldBy = holdTicket.onHoldBy;
        this.closedBy = holdTicket.closedBy;
        this.comments = holdTicket.comments;
        this.dispositionComments = holdTicket.dispositionComments;
        this.reverificationRequired = holdTicket.reverificationRequired;
        this.responsibleForAction = holdTicket.responsibleForAction;
        this.dispositionBy = holdTicket.dispositionBy;
        this.cparNum = holdTicket.cparNum;
        this.readonly = true;

        this.dispositionDate = this.dateToString(holdTicket.dispositionDate);
        this.closedDate = this.dateToString(holdTicket.closedDate);
        this.holdDate = this.dateToString(holdTicket.holdDate);
        this.productionDate = this.dateToString(holdTicket.productionDate);
    }

    calculateDaysOpen(): number {

        if (this.holdDate) {

            let today = new Date().toISOString().split('T');
            let end = new Date(this.holdDate).toISOString().split('T');
            let momentToday = moment(today[0], 'YYYY-MM-DD');
            let momentEnd = moment(end[0], 'YYYY-MM-DD');
            return Math.abs(momentEnd.diff(momentToday, "days"));

        } else {
            return 0;
        }
    }

    dateToString(dateString): string {

        if (dateString) {

            let date = new Date(dateString).toISOString().split('T');
            return date[0];

        } else {
            return "Date Not Set";
        }
    }

    stringToDate(dateString): Date {

        return new Date(dateString);

    }

    toJSON() {
        return {
            holdTicketId: this.holdTicketId,
            parentId: this.parentId,
            dispositionId: this.dispositionId,
            defectId: this.defectId,
            plantCode: this.plantCode,
            statusId: this. statusId,
            prodOrderNum: this.prodOrderNum,
            prodOrderQty: this.prodOrderQty,
            salesOrderNum: this.salesOrderNum,
            returnAuthNum: this.returnAuthNum,
            purchaseOrderNum: this.purchaseOrderNum,
            materialDesc: this.materialDesc,
            materialNumber: this.materialNumber,
            lineItem: this.lineItem,
            optionNumber: this.optionNumber,
            color: this.color,
            workCenter: this.workCenter,
            producedByShift: this.producedByShift,
            qtyOnHold: this.qtyOnHold,
            totalSkids: this.totalSkids,
            skidNumbersOnHold: this.skidNumbersOnHold,
            onHoldBy: this.onHoldBy,
            closedBy: this.closedBy,
            comments: this.comments,
            dispositionComments: this.dispositionComments,
            reverificationRequired: this.reverificationRequired,
            responsibleForAction: this.responsibleForAction,
            dispositionBy: this.dispositionBy,
            cparNum: this.cparNum,
            dispositionDate: this.stringToDate(this.dispositionDate),
            closedDate: this.stringToDate(this.closedDate),
            holdDate: this.stringToDate(this.holdDate),
            productionDate: this.stringToDate(this.productionDate)
        }
  }
}

File: src/main/java/com/orbis/holdwip/domain/hold/entity/HoldTicket.java
package com.orbis.holdwip.domain.hold.entity;

import javax.persistence.*;
import java.util.Date;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.orbis.holdwip.domain.hold.Defect;

@Entity
public class HoldTicket {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long holdTicketId;
    private Long parentId;
    private int dispositionId;
    private int defectId;
    private String plantCode;
    private int statusId;
    private String prodOrderNum;
    private String prodOrderQty;
    private String salesOrderNum;
    private String returnAuthNum;
    private String purchaseOrderNum;
    private String materialDesc;
    private String materialNumber;
    private String lineItem;
    private String optionNumber;
    private String color;
    private String workCenter;
    private String producedByShift;
    private String qtyOnHold;
    private String totalSkids;
    private String skidNumbersOnHold;
    private String onHoldBy;
    private String closedBy;
    private String comments;
    private String dispositionComments;
    private String reverificationRequired;
    private String responsibleForAction;
    private String dispositionBy;
    private String cparNum;
    private Date productionDate;
    private Date holdDate;
    private Date dispositionDate;
    private Date closedDate;

    @OneToMany(cascade = CascadeType.ALL, mappedBy="holdTicket")
    @JsonManagedReference
    private Set<HoldTicketAttachment> attachments;

    @OneToMany(cascade = CascadeType.ALL, mappedBy="holdTicket")
    private Set<HoldActionItem> actionItems;

    @OneToMany(cascade = CascadeType.ALL, mappedBy="holdTicket")
    private Set<HoldQiDetail> qiDetails;

    @Transient
    int labor;

    public Long getHoldTicketId() {
        return holdTicketId;
    }

    public void setHoldTicketId(Long holdTicketId) {
        this.holdTicketId = holdTicketId;
    }

    public Long getParentId() {
        return parentId;
    }

    public void setParentId(Long parentId) {
        this.parentId = parentId;
    }

    public int getDispositionId() {
        return dispositionId;
    }

    public void setDispositionId(int dispositionId) {
        this.dispositionId = dispositionId;
    }

    public int getDefectId() {
        return defectId;
    }

    public void setDefectId(int defectId) {
        this.defectId = defectId;
    }

    public String getPlantCode() {
        return plantCode;
    }

    public void setPlantCode(String plantCode) {
        this.plantCode = plantCode;
    }

    public int getStatusId() {
        return statusId;
    }

    public void setStatusId(int statusId) {
        this.statusId = statusId;
    }

    public String getProdOrderNum() {
        return prodOrderNum;
    }

    public void setProdOrderNum(String prodOrderNum) {
        this.prodOrderNum = prodOrderNum;
    }

    public String getProdOrderQty() {
        return prodOrderQty;
    }

    public void setProdOrderQty(String prodOrderQty) {
        this.prodOrderQty = prodOrderQty;
    }

    public String getSalesOrderNum() {
        return salesOrderNum;
    }

    public void setSalesOrderNum(String salesOrderNum) {
        this.salesOrderNum = salesOrderNum;
    }

    public String getReturnAuthNum() {
        return returnAuthNum;
    }

    public void setReturnAuthNum(String returnAuthNum) {
        this.returnAuthNum = returnAuthNum;
    }

    public String getPurchaseOrderNum() {
        return purchaseOrderNum;
    }

    public void setPurchaseOrderNum(String purchaseOrderNum) {
        this.purchaseOrderNum = purchaseOrderNum;
    }

    public String getMaterialDesc() {
        return materialDesc;
    }

    public void setMaterialDesc(String materialDesc) {
        this.materialDesc = materialDesc;
    }

    public String getMaterialNumber() {
        return materialNumber;
    }

    public void setMaterialNumber(String materialNumber) {
        this.materialNumber = materialNumber;
    }

    public String getLineItem() {
        return lineItem;
    }

    public void setLineItem(String lineItem) {
        this.lineItem = lineItem;
    }

    public String getOptionNumber() {
        return optionNumber;
    }

    public void setOptionNumber(String optionNumber) {
        this.optionNumber = optionNumber;
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }

    public String getWorkCenter() {
        return workCenter;
    }

    public void setWorkCenter(String workCenter) {
        this.workCenter = workCenter;
    }

    public String getProducedByShift() {
        return producedByShift;
    }

    public void setProducedByShift(String producedByShift) {
        this.producedByShift = producedByShift;
    }

    public String getQtyOnHold() {
        return qtyOnHold;
    }

    public void setQtyOnHold(String qtyOnHold) {
        this.qtyOnHold = qtyOnHold;
    }

    public String getTotalSkids() {
        return totalSkids;
    }

    public void setTotalSkids(String totalSkids) {
        this.totalSkids = totalSkids;
    }

    public String getSkidNumbersOnHold() {
        return skidNumbersOnHold;
    }

    public void setSkidNumbersOnHold(String skidNumbersOnHold) {
        this.skidNumbersOnHold = skidNumbersOnHold;
    }

    public String getOnHoldBy() {
        return onHoldBy;
    }

    public void setOnHoldBy(String onHoldBy) {
        this.onHoldBy = onHoldBy;
    }

    public String getClosedBy() {
        return closedBy;
    }

    public void setClosedBy(String closedBy) {
        this.closedBy = closedBy;
    }

    public String getComments() {
        return comments;
    }

    public void setComments(String comments) {
        this.comments = comments;
    }

    public String getDispositionComments() {
        return dispositionComments;
    }

    public void setDispositionComments(String dispositionComments) {
        this.dispositionComments = dispositionComments;
    }

    public String getReverificationRequired() {
        return reverificationRequired;
    }

    public void setReverificationRequired(String reverificationRequired) {
        this.reverificationRequired = reverificationRequired;
    }

    public String getResponsibleForAction() {
        return responsibleForAction;
    }

    public void setResponsibleForAction(String responsibleForAction) {
        this.responsibleForAction = responsibleForAction;
    }

    public String getDispositionBy() {
        return dispositionBy;
    }

    public void setDispositionBy(String dispositionBy) {
        this.dispositionBy = dispositionBy;
    }

    public String getCparNum() {
        return cparNum;
    }

    public void setCparNum(String cparNum) {
        this.cparNum = cparNum;
    }

    public Date getProductionDate() {
        return productionDate;
    }

    public void setProductionDate(Date productionDate) {
        this.productionDate = productionDate;
    }

    public Date getHoldDate() {
        return holdDate;
    }

    public void setHoldDate(Date holdDate) {
        this.holdDate = holdDate;
    }

    public Date getDispositionDate() {
        return dispositionDate;
    }

    public void setDispositionDate(Date dispositionDate) {
        this.dispositionDate = dispositionDate;
    }

    public Date getClosedDate() {
        return closedDate;
    }

    public void setClosedDate(Date closedDate) {
        this.closedDate = closedDate;
    }

    //@JsonIgnore
    public Set<HoldTicketAttachment> getAttachments() {
        return attachments;
    }

    public void setAttachments(Set<HoldTicketAttachment> attachments) {
        this.attachments = attachments;
    }

    public Set<HoldActionItem> getActionItems() {
        return actionItems;
    }

    public void setActionItems(Set<HoldActionItem> actionItems) {
        this.actionItems = actionItems;
    }

    public Set<HoldQiDetail> getQiDetails() {
        return qiDetails;
    }

    public void setQiDetails(Set<HoldQiDetail> qiDetails) {
        this.qiDetails = qiDetails;
    }

    public int getLabor() {
        return labor;
    }

    public void setLabor(int labor) {
        this.labor = labor;
    }

    //get defect description
    public String getDefectDesc (int defectId){
        defectId = this.defectId;
        //get the defect description based on the ID
        return Defect.getDefectFromId(defectId).getDesc();
    }
}
